[FEAT/#134] core / PhotoPicker 구현!

shinythinking opened the pull request:
- closed #134 
## *📍 Work Description*

- 버전별 권한 대응
  - 두 부분으로 이루어집니다.
  - 하나(`PermissionCheck`)는 기기의 버전에 맞는 권한을 반환하는 클래스입니다. 
  - 다른 하나(`PermissionHandler`)는 권한이 있는지 파악하여, 권한을 요구하는 컴포저블 함수입니다.
-  카메라
   - 카메라로 사진을 찍으면 로컬 저장소에 저장됩니다.
   - 이는 포토피커에 반영되고, 선택됩니다. 이 때, 다시 저장소에서 가져오는 것이 아니고, 저장되었다고 가정하고 표현합니다. 
- 선택할 수 있는 사진 개수를 인자로 받습니다.
  - 선택한 사진의 개수가 topBar에 반영됩니다. 
  - max를 넘기면 사진도 못 찍고, 사진 선택도 불가능합니다.
- url 대신 PhotoInfo를 사용한 이유
  -  PhotoInfo에는 url을 포함하여, 사진에 대한 정보를 가지고 있습니다. 사진의 메타 정보를 활용해야 하는 피처가 있을 수 있기 때문에 photoInfo를 도입했습니다. 
- 오류 대응
  -  한번에 하나의 권한만 물어보는 문제
  - 카메라로 2번 이상 찍으면 여러 장이 저장되는 문제
  - 카메라로 찍은 사진이 거꾸로 저장되는 문제
- 권한 취소 대응
  - 권한 요청을 취소하면 Dialog를 띄우고 설정 화면으로 이동할 수 있게 합니다.
  - 카메라와 저장소 권한 모두를 갖지 않으면 앱 실행이 어렵도록 했습니다.
  

## *📸 Screenshot*
### 카메라 없을 때,
<img src="https://github.com/user-attachments/assets/f8439784-bc5e-4604-bfcb-f5a1b6643f62" width=270 />

### 카메라 있을 때,
<img src="https://github.com/user-attachments/assets/47fc285c-0dec-4786-b2b0-8f266f17d08e" width=270 />

### 피드백 반영

#### 체크 표시가 잘 보이지 않던 문제 해결
<img src="https://github.com/user-attachments/assets/6def83f9-c877-4224-8805-90cdf466b52a" width=270 />
==================>
<img src="https://github.com/user-attachments/assets/9a5d1997-9e49-4270-bfa7-e65f6538e7aa" width=270 />

#### 권한 취소 대응

<img src="https://github.com/user-attachments/assets/7fea5ea8-9307-4e9a-8715-0d83f4e20f8f" width=270 />


- 권한 dialog에서 **확인**을 누르면 설정화면으로 이동합니다. 이후 다시 앱으로 돌아가면 사진을 로드합니다.



<img src="https://github.com/user-attachments/assets/300bd12f-9e26-4e04-afdb-3c98f3d9725c" width=270 />


- **취소**하면 dialog를 계속 띄웁니다. 모두 승인해야 사용이 가능하도록 설계했습니다.



## *📢 To Reviewers*
- 믿고 기다려주신 덕분에, 오랜만에 재미있게 프로그래밍했습니다. 
- 카메라 아이콘 배경에 어떤 색을 넣으면 좋을지 고민입니다. 
- 우선은 카카오톡의 ui를 기준으로 만들었습니다.
- Recomposition 대응을 완전히 하지 못 했습니다. 리팩토링 기간에 persistentList 도입하도록 하겠습니다.
- 적용하는 방법과 관련해서는 #5cc8b2c08564d791441b1c2b89683b6c1aeec1bc 을 확인하시거나 저한테 언제든 편하게 물어보시면 됩니다.
- 이후에 작업할 목록
  - 완료 비활성화시, 색상 변경
## ⏲️Time

    - 16 h

shinythinking committed:
chore/#134 : 체크 아이콘 추가 및 체크박스 컬러 추가

shinythinking committed:
feat/#134 : PhotoInfo 클래스 구성

shinythinking committed:
feat/#134 : 저장소 접근 권한 관리 구현

- 안드로이드 버전별 저장소에 접근하기 위해 필요한 권한이 상이합니다. 이에 대응하기 위해서 PhotoPermissionCheck 클래스를 만들었습니다.
- PhotoPermissionHandler는 권한이 있는지 확인한 후, 권한 요청을 진행합니다.

shinythinking committed:
feat/#134 : 카메라 접근 권한 관리 구현

shinythinking committed:
feat/#134 : Manifest 파일에 버전별 권한 요청

shinythinking committed:
feat/#134 : 저장소에 접근하여 쿼리를 이용해서 사진을 가져오는 객체 구현

shinythinking committed:
feat/#134 : 찍은 사진을 저장하는 객체

- 이름에는 찍은 시간을 담았습니다.
- jpeg로 저장합니다.

shinythinking committed:
chore/#134 : core:designsystem -> core:ui 의존성 삭제

shinythinking committed:
feat/#134 : 필요 없는 parcelable 삭제

shinythinking committed:
chore/#134 : 카메라 이미지 추가

shinythinking committed:
feat/#134 : MapisodePhotoPicker 구현

- 크게 두 부분으로 이루어져있습니다.
- MapisodePhotoPicker는 권한 관련 처리를 우선으로 합니다.
- 실제 보여주는 화면은 PhotoPicker입니다.

shinythinking committed:
chore/#134 : 디렉토리 정리

shinythinking committed:
chore/#134 : coil 의존성 추가

shinythinking committed:
feat/#134 : MapisodePhotoPicker 추가!

1. isPhotoPickerClicked : 포토피커를 보여줄지 결정하는 상태 변수
2. 분기문을 이용해서 photopicker 화면과 기존 화면 분리
3. photopicker 사용!

shinythinking committed:
feat/#134 : profileUrl 변경 intent 추가

shinythinking committed:
feat/#134 : profileUrl 변경 처리 함수 추가

shinythinking committed:
feat/#134 : profileUrl 변경 관련 변수 추가

shinythinking committed:
fix/#134 : 권한을 한번에 하나만 물어보는 오류 해결

shinythinking committed:
fix/#134 : remember -> rememberSaveable

shinythinking committed:
feat/#134 : photoInfo 클래스 parcelable로 변경

shinythinking committed:
fix/#134 : remember -> rememberSaveable for snapshotStateList

- snapshotStateList를 rememberSaveable로 저장하기 위한 ext 추가

shinythinking committed:
fix/#134 : 사진 두 번 찍으면 3개가 저장되는 오류 해결

- 이전에는 사진 찍은 것을 addedPhotos로 저장해서 SelectedPhotos에 저장.
- 이러면 addedPhotos에 이전에 찍었던 사진이 있어서 중복 저장.
- 이걸 방지하고자 찍은 사진 하나만 넘겨주어 저장

shinythinking committed:
fix/#134 : 찍은 사진들이 거꾸로 저장되는 문제 해결

shinythinking committed:
fix/#134 : ktlint 대응

TaewoongR commented on core/ui/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,18 @@
+
+	<uses-permission android:name="android.permission.CAMERA" />
+
+	<!-- Android 13 이상 (API 33 이상) -->
+	<uses-permission android:name="android.permission.READ_MEDIA_IMAGES" />
+
+	<!-- Android 10-12 (API 29-32) -->
+	<uses-permission
+		android:name="android.permission.READ_EXTERNAL_STORAGE"
+		android:maxSdkVersion="32" />
+
+	<!-- Android 9 이하 (API 28 이하) -->
+	<uses-permission
+		android:name="android.permission.WRITE_EXTERNAL_STORAGE"
+		android:maxSdkVersion="28" />
```
버전 대응 멋있습니다~

shinythink replied:
감사합니다 ㅎㅎ

TaewoongR commented on core/ui/build.gradle.kts:
```diff
@@ -1,7 +1,14 @@
 plugins {
 	alias(libs.plugins.mapisode.android.compose)
+	id("kotlin-parcelize")
```
아무런 플러그인 id 명시없이 곧바로 쓸 수 있도록 되어있는 플러그인이네요
[[Parcelable implementation generator]](https://developer.android.com/kotlin/parcelize)
공식 가이드에서도 곧바로 쓰긴하네요ㄷㄷ

대신 버전 카탈로그 일관성 지키려면 아래처럼 해보면 좋을거 같아요
[[Kotlin parcelize plugin in Version Catalog]](https://medium.com/@htooko1101/kotlin-parcelize-plugin-in-version-catalog-d5ff2e1752d9)

버전은 코틀린 버전 그대로 따라가네요.
[[plugin/org.jetbrains.kotlin.plugin.parcelize]](https://plugins.gradle.org/plugin/org.jetbrains.kotlin.plugin.parcelize/2.0.21)

shinythink replied:
와...ㅠ 진짜 감사합니다!!!!!!!

TaewoongR commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/PhotoSaver.kt:
```diff
@@ -0,0 +1,55 @@
+		val imageFileName = "JPEG_$timestamp"
+
+		val contentValues = ContentValues().apply {
+			put(MediaStore.MediaColumns.DISPLAY_NAME, imageFileName)
+			put(MediaStore.MediaColumns.MIME_TYPE, "image/jpeg")
+
+			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+				put(MediaStore.MediaColumns.RELATIVE_PATH, Environment.DIRECTORY_PICTURES)
+				put(MediaStore.Images.Media.IS_PENDING, 1)
+			}
```
만약에 Build.VERSION_CODES.Q 버전보다 낮은 경우에는  위의 명령어는 실행되지 않는 건가요?

shinythink replied:
버전이 낮은 경우에는 상대 경로로 접근을 하지 않고 절대 경로로 접근한다고 합니다. 그래서 저 코드가 필요가 없습니다. 보안상 더 까다로워졌습니다. Scoped Storage 찾아보시면 깊게 이해하실 수 있으실 듯 합니다!

TaewoongR commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/PhotoSaver.kt:
```diff
@@ -0,0 +1,55 @@
+			contentValues,
+		) ?: return null
+
+		try {
+			contentResolver.openOutputStream(imageUri)?.use { outputStream ->
+				bitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream)
+			}
+
+			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+				contentValues.clear()
+				contentValues.put(MediaStore.Images.Media.IS_PENDING, 0)
+				contentResolver.update(imageUri, contentValues, null, null)
+			}
+
+			return imageUri
+		} catch (e: IOException) {
+			contentResolver.delete(imageUri, null, null)
+			return null
+		}
```
요 부분에서 이미지 저장 성공하면 Uri를 반환하는거죠?

shinythink replied:
넵 그렇습니다! 현재 찍은 사진의 uri를 반환합니다!

haeti-dev replied:
실패했을 때 처리 부탁드립니다

TaewoongR commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/component/CircleWithCheck.kt:
```diff
@@ -0,0 +1,40 @@
+import com.boostcamp.mapisode.designsystem.compose.MapisodeIcon
+import com.boostcamp.mapisode.designsystem.theme.MapisodeTheme
+
+@Composable
+fun CircleWithCheck(
+	checked: Boolean,
+	modifier: Modifier = Modifier,
+) {
+	Box(
+		modifier = modifier
+			.padding(4.dp)
+			.size(28.dp)
+			.background(
+				color = MapisodeTheme.colorScheme.checkboxBackground,
+				shape = CircleShape,
+			)
+			.border(
+				width = 1.dp,
+				color = MapisodeTheme.colorScheme.checkboxStroke,
+				shape = CircleShape,
+			),
+		contentAlignment = Alignment.Center,
+	) {
+		if (checked) {
+			MapisodeIcon(id = com.boostcamp.mapisode.designsystem.R.drawable.ic_check)
+		}
+	}
+}
```
백그라운드 투명에 border만 주고 체크되면 아이콘 띄워주는 방식으로 처리했네요.
저도 이 부분 어떻게 했는지 궁금했어요.

shinythink replied:
저도 고민을 좀 했는데 직관적인 방식으로 했습니다 ㅎㅎ

TaewoongR commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/component/PhotoItem.kt:
```diff
@@ -0,0 +1,49 @@
+			.clickable(onClick = onPhotoClick),
+	) {
+		Box(modifier = Modifier.fillMaxSize()) {
+			Image(
+				painter = rememberAsyncImagePainter(
+					ImageRequest.Builder(LocalContext.current).data(data = photo.uri).build(),
+				),
+				contentDescription = "Photo",
+				contentScale = ContentScale.Crop,
+				modifier = Modifier.fillMaxSize(),
+			)
```
이 함수는 잘 안써봐서 몰랐는데 이런식으로 로컬 사진을 불러오는군요

shinythink replied:
coil에서 이렇게 사용하는 예시가 있더라구요.
전 처음 써봤습니다.

TaewoongR commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/model/PhotoInfo.kt:
```diff
@@ -0,0 +1,8 @@
+import kotlinx.parcelize.Parcelize
+import java.util.Date
+
+@Parcelize
+data class PhotoInfo(val uri: String, val dateTaken: Date?) : Parcelable
```
이 방식이 안드로이드 내부에서 데이터를 전달하는데 Serialize보다 안정적이라 하네요.

shinythink replied:
속도도 더 빠르다고 하더라구요 ㅎㅎ

TaewoongR commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/component/CircleWithCheck.kt:
```diff
@@ -0,0 +1,40 @@
+			.padding(4.dp)
+			.size(28.dp)
+			.background(
+				color = MapisodeTheme.colorScheme.checkboxBackground,
```
체크될 때 배경색이 투명인 경우 체크표시 색상과 이미지 색상이 구분이 안될 수도 있겠네요

shinythink replied:
좋은 피드백 감사합니다!! 수정하도록 하겠습니다!!!

TaewoongR commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/MapisodePhotoPicker.kt:
```diff
@@ -0,0 +1,179 @@
+import java.time.Instant
+import java.util.Date
+
+@Composable
+fun MapisodePhotoPicker(
+	numOfPhoto: Int,
+	onPhotoSelected: (List<PhotoInfo>) -> Unit,
+	onBackPressed: () -> Unit,
+	modifier: Modifier = Modifier.fillMaxSize(),
+	isCameraNeeded: Boolean = true,
+) {
```
<img src="https://github.com/user-attachments/assets/2fdeaf1d-9ca7-49d8-b3c4-c1242d009b84" width=600>

함수 파라미터 modifier 기본값은 항상 Modifier 만 있어야해요.
아래에 modifier가 쓰이는 곳에서 fillMaxSize() 적용해주시면 될거 같아요.

shinythink replied:
오!! 놓치고 있었네요!! 감사합니다! 꼼꼼한 피드백 감동입니다 🤩

TaewoongR commented on feature/login/src/main/java/com/boostcamp/mapisode/login/SignUpScreen.kt:
```diff
@@ -10,97 +11,132 @@ import androidx.compose.foundation.layout.height
+			isCameraNeeded = false,
+		)
+	} else {
+		MapisodeScaffold(
```
이렇게 했을 때 아쉬운 점이 하나가 있어요.
디바이스 뒤로가기를 눌렀을 때 전 화면이 아닌 앱 밖으로 나가져요.

shinythink replied:
정성 가득한 피드백 감사합니다!
이 부분 해결했습니다! 
덕분에 ux가 많이 향상된 것 같습니다!

TaewoongR commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/permissionHandler/PermissionHandler.kt:
```diff
@@ -0,0 +1,44 @@
+		mutableStateOf(PermissionCheck(context).hasPermissions(isCameraNeeded))
+	}
+
+	val launcher = rememberLauncherForActivityResult(
+		contract = ActivityResultContracts.RequestMultiplePermissions(),
+	) { permissions ->
+		val allGranted = permissions.values.all { it }
+		hasPermissions = allGranted
+
+		if (allGranted) {
+			onPermissionsGranted()
+		} else {
+			onPermissionsDenied()
+		}
+	}
+
+	LaunchedEffect(hasPermissions) {
+		if (!hasPermissions) {
+			launcher.launch(PermissionCheck.getRequiredPermissions(isCameraNeeded))
+		} else {
+			onPermissionsGranted()
+		}
+	}
```
여기서 허용권한 선택 창이 뜨는건가요?

shinythink replied:
예 그렇습니다! 
띄우기 전에 카메라에 대한 권한도 필요한지 물어보는 로직이 PermissionCheck 안에 있습니다.

TaewoongR reviewed (approved):
기본 Photo picker 를 사용하면 굉장히 쉬운데 직접 구현하는건 어렵네요.
코드 자체도 화면 UI 구현보다 어렵네요.

shinythinking committed:
fix/#134 : 뒤로가기 버튼 클릭시, 앱이 꺼지던 문제 해결

shinythinking committed:
fix/#134 : modifier의 기본값 수정

shinythinking committed:
fix/#134 : 체크표시 잘 보이게 수정

shinythinking committed:
refactor/#134 : id(parcelize) -> alias(parcelize) 방식으로 변경

haeti-dev commented on core/ui/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest xmlns:android="http://schemas.android.com/apk/res/android">
+
+	<uses-permission android:name="android.permission.CAMERA" />
+
+	<!-- Android 13 이상 (API 33 이상) -->
+	<uses-permission android:name="android.permission.READ_MEDIA_IMAGES" />
+
+	<!-- Android 10-12 (API 29-32) -->
+	<uses-permission
+		android:name="android.permission.READ_EXTERNAL_STORAGE"
+		android:maxSdkVersion="32" />
+
+	<!-- Android 9 이하 (API 28 이하) -->
+	<uses-permission
+		android:name="android.permission.WRITE_EXTERNAL_STORAGE"
+		android:maxSdkVersion="28" />
```
앱 추출과 출시를 위해서라면 app 모듈에도 추가해야 합니다~

shinythink replied:
아 그렇군요! 
감사합니다!!!!!

haeti-dev commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/MapisodePhotoPicker.kt:
```diff
@@ -0,0 +1,179 @@
+
+	val photoLoader = remember(context) { PhotoLoader(context) }
+
+	LaunchedEffect(isPermissionsGranted) {
+		if (isPermissionsGranted) {
+			try {
+				photos = photoLoader.loadPhotos(100)
+			} catch (e: Exception) {
+				e.printStackTrace()
+			}
+		}
+	}
```
권한을 거부하면 앱이 꺼지나요? 권한을 거부했을때 다이얼로그나 토스트로 앱에서 이 기능이 필수적이라는 것을 알려줄지, 프로필 사진을 등록하지 않고도 앱을 사용할 수 있게 유도할건지 반영해주시면 좋을 것 같습니다~

haeti-dev replied:
그리고 키를 isPermissionsGranted로 한 이유가 있나요? 지금 권한을 재요청하지 않는데, Unit이랑 뭐가 다른지 모르겠네요

shinythink replied:
이 부분을 고려하다가 넘어갔었네요 ㅠㅠ 말씀해주셔서 감사합니다.
현재는 거부하면 빈 화면을 보여줍니다.
지금 생각으로는 권한을 거부하면
1. 권한을 설정하는 방법을 토스트로 알려주고
2. 프로필 이미지 설정에 한해서는, 사진을 선택하지 않고 완료를 누르면 기본이미지를 배정하고
3. 빈화면을 보여주는 대신, 닉네임을 설정하는 화면을 보여주는 것으로 하겠습니다.

shinythink replied:
키를 isPermissionGranted로 한 이유는 권한을 설정하고 다시 photopicker를 눌렀을 때, 사진을 로드하기 위함입니다. 좀 더 좋은 방법이 있을 것 같은데, 제 한계입니다 ㅠㅠ 좀 더 찾아보고 보완할 수 있으면 해보겠습니다!

haeti-dev commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/MutableStateListExt.kt:
```diff
@@ -0,0 +1,18 @@
+import androidx.compose.runtime.snapshots.SnapshotStateList
+import androidx.compose.runtime.toMutableStateList
+
+@Composable
+fun <T : Any> rememberMutableStateListOf(vararg elements: T): SnapshotStateList<T> =
+	rememberSaveable(saver = snapshotStateListSaver()) {
+		elements.toList().toMutableStateList()
+	}
+
+private fun <T : Any> snapshotStateListSaver() = listSaver<SnapshotStateList<T>, T>(
+	save = { stateList -> stateList.toList() },
+	restore = { it.toMutableStateList() },
+)
```
왜 필요하고 어떤 동작을 하는지 설명해주세요

shinythink replied:
현재 코드에서는 선택된 사진을 저장하는데 MutableStateList를 사용합니다. 이걸 configuration 동안에 보존하고 싶은데, MutableStateList가 SnashotStateList여서 rememberSaveable 사용이 불가능합니다. 이를 해결하기 위해 도입했습니다. [링크](https://stackoverflow.com/questions/68885154/using-remembersaveable-with-mutablestatelistof)를 참고했습니다.

haeti-dev commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/PhotoLoader.kt:
```diff
@@ -0,0 +1,62 @@
+
+class PhotoLoader(private val context: Context) {
+
+	suspend fun loadPhotos(
+		limit: Int = 100,
+		offset: Int = 0,
+	): List<PhotoInfo> = withContext(Dispatchers.IO) {
```
페이징으로 리펙토링 하시면 좋을 것 같습니다

haeti-dev replied:
limit과 offset을 사용하고 있지 않은데 로직 추가해주시기 바랍니다

shinythink replied:
원래는 sql의 limit과 offset 구문을 생각하고 추가했었는데, 알고보니 안드에서는 사용이 불가능한 구문이었습니다. 삭제한다는 걸 깜빡했습니다. 우선 남겨두고 리팩토링하는데 활용해보겠습니다!

![Screenshot 2024-12-01 at 12 27 40 AM](https://github.com/user-attachments/assets/2862c0c2-803e-40b9-b74c-dfcb384e2c7c)

 꼼꼼한 피드백 진심으로 감사합니다!

haeti-dev commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/PhotoSaver.kt:
```diff
@@ -0,0 +1,55 @@
+
+		try {
+			contentResolver.openOutputStream(imageUri)?.use { outputStream ->
+				bitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream)
```
100은 뭘 의미하나요?

haeti-dev replied:
https://kotlinlang.org/docs/functions.html#named-arguments

전체적으로 이름있는 argunment 사용해주시면 좋을 것 같아요~

shinythink replied:
100은 품질(quality)를 나타냅니다. 현재는 무손실입니다. 이 부분은 함께 이야기해봐도 좋을 것 같습니다. 영수님의 발표에 따르면 80으로 해도 좋을 것 같습니다.

shinythink replied:
이 함수가 자바 코드로 이루어져 있어서 이름있는 argument 사용을 못했습니다. 
![Screenshot 2024-12-01 at 12 31 54 AM](https://github.com/user-attachments/assets/3ee312c9-d056-4bd8-948e-699ee4da8755)

haeti-dev commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/component/PhotoItem.kt:
```diff
@@ -0,0 +1,49 @@
+			.clickable(onClick = onPhotoClick),
+	) {
+		Box(modifier = Modifier.fillMaxSize()) {
+			Image(
+				painter = rememberAsyncImagePainter(
+					ImageRequest.Builder(LocalContext.current).data(data = photo.uri).build(),
+				),
+				contentDescription = "Photo",
+				contentScale = ContentScale.Crop,
+				modifier = Modifier.fillMaxSize(),
+			)
```
이 부분은 화면에 많은 사진들을 로딩하는 부분이니 placeholder나 error 이미지를 추가하면 좋을 것 같아요

shinythink replied:
넵 알겠습니다. 
추가했습니다!
조언 감사합니다!!

haeti-dev commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/permissionHandler/PermissionCheck.kt:
```diff
@@ -0,0 +1,38 @@
+
+	}
+
+	fun hasPermissions(isCameraNeeded: Boolean): Boolean =
```
`arePermissionsGranted` 라는 함수명이 더 적합할 것 같아요

shinythink replied:
동의합니다!! 수정하겠습니다!!

haeti-dev commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/MapisodePhotoPicker.kt:
```diff
@@ -0,0 +1,179 @@
+				item {
+					CameraItem(
+						onCameraClick = {
+							if (selectedPhotos.size < numOfPhoto) {
+								onCameraClick()
+							}
```
최대 선택개수보다 많이 선택했을 때 토스트로 "최대 x개의 사진만 선택할 수 있습니다" 라는 UX를 제공하면 좋을 것 같아요. TopBar로만 정보를 제공하는게 아쉽습니다.

shinythink replied:
추가하면 ux 상으로 훨씬 좋아질 것 같습니다! 
추가했습니다!!
😊

haeti-dev commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/PhotoSaver.kt:
```diff
@@ -0,0 +1,55 @@
+import java.util.Locale
+
+object PhotoSaver {
+	fun savePhoto(context: Context, bitmap: Bitmap?): Uri? {
```
메인 스레드에서 호출될 경우 UI 스레드가 터져서 ANR이 일어날 수 있을 것 같아요. suspend로 변경해주시고 호출부에서도 coroutineScope로 감싸주세요

shinythinking replied:
그 부분을 고려하지 못했네요 ㅠ 
진짜 많이 배웁니다! 
항상 감사해요~😊

shinythink replied:
혹시 ANR이 어디서 터질까요??
어디에 CoroutineScope를 적용해야 할지 고민입니다.

코루틴에 대한 이해가 부족합니다 ㅠㅠ

Ameri-Kano replied:
이 함수가 사진을 저장하는 로직 자체라면 사진을 저장하는 함수 (save 같은) 부분에서 Dispatchers.IO 스코프로 실행하면 될 것 같습니다

haeti-dev commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/permissionHandler/PermissionHandler.kt:
```diff
@@ -0,0 +1,44 @@
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.platform.LocalContext
+
+@Composable
+fun PermissionHandler(
+	isCameraNeeded: Boolean,
+	onPermissionsGranted: () -> Unit,
+	onPermissionsDenied: () -> Unit,
+) {
+	val context = LocalContext.current
+	var hasPermissions by rememberSaveable {
+		mutableStateOf(PermissionCheck(context).hasPermissions(isCameraNeeded))
+	}
```
`hasPermissions` 상태와 `onPermissionsGranted`, `onPermissionsDenied` 콜백의 역할이 겹칩니다. `hasPermissions` 상태를 상위 컴포저블로 이동하고, `PermissionHandler`는 권한 요청만 담당하도록 단순화할 수 있을 것 같아요

shinythink replied:
오랜만에 벽을 느끼게 되는 리뷰입니다..ㅠ
오래 걸릴 것 같아서 우선 머지하고 천천히 고치도록 하겠습니다.
좋은 리뷰 감동입니다.

haeti-dev commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/MapisodePhotoPicker.kt:
```diff
@@ -0,0 +1,179 @@
+	onCameraClick: () -> Unit,
+	modifier: Modifier = Modifier,
+) {
+	val selectedPhotos = rememberMutableStateListOf<PhotoInfo>()
+	if (addedPhoto != null) selectedPhotos.add(addedPhoto)
+
```
리펙토링때 이 부분 리컴포지션 신경써주세요~

shinythink replied:
넵 알겠습니다! 
리팩토링 주간에는 리컴포지션 제대로 다뤄볼게요!

haeti-dev commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/MapisodePhotoPicker.kt:
```diff
@@ -0,0 +1,179 @@
+import java.time.Instant
+import java.util.Date
+
+@Composable
+fun MapisodePhotoPicker(
```
너무 많은 책임을 가지고 있어요. 사진로드, 권한 처리, 카메라 실행을 적절하게 분리할 수 있어보입니다

shinythink replied:
이 부분 리팩토링은 리팩토링 주간에 진행하도록 하겠습니다!
머리를 좀 더 써야할 것 같습니다 ㅠㅠ

haeti-dev reviewed (approved):
동역님이 필요하신 코드니 승인 먼저 드립니다. 
전체적으로 에러 헨들링, 리컴포지션 관리가 더 되어야 할 것 같고, 코드 간의 스타일 불일치가 존재하네요 리뷰 바탕으로 적절하게 수정해주시면 감사하겠습니다.
너무 고생하셨어요~

shinythinking committed:
fix/#134 : app 모듈에도 필요 권한 명시

* 카메라의 하드웨어 기능 선언 추가

shinythinking committed:
refactor/#134 : hasPermissions -> arePermissionsGranted 이름 변경

shinythinking committed:
refactor/#134 : 최대 허용 사진 개수 초과시, Toast를 띄움

shinythinking committed:
feat/#134 : 권한 설정 대응

- 권한 요청을 취소하면 Dialog를 띄우고 설정 화면으로 이동할 수 있게 합니다.
-  카메라와 저장소 권한 모두를 갖지 않으면 앱 실행이 어렵도록 했습니다.

shinythinking committed:
feat/#134 : 사진 저장 실패 대응

shinythinking committed:
feat/#134 : 이미지를 불러올 때, error이미지와 placeholder 이미지 추가

Ameri-Kano commented on build.gradle.kts:
```diff
@@ -9,6 +9,7 @@ plugins {
 	alias(libs.plugins.android.library) apply false
 	alias(libs.plugins.ktlint) apply false
 	alias(libs.plugins.google.services) apply false
+	alias(libs.plugins.kotlin.parcelize) apply false
```
별거 아니죠?

shinythink replied:
네 ㅎㅎ 생각보다 뭐 없네요 ㅋㅋㅋ 감사합니다~

Ameri-Kano commented on core/ui/src/main/kotlin/com/boostcamp/mapisode/ui/photopicker/PhotoLoader.kt:
```diff
@@ -0,0 +1,62 @@
+		val sortOrder = "${MediaStore.Images.Media.DATE_TAKEN} DESC"
+		val collection = MediaStore.Images.Media.EXTERNAL_CONTENT_URI
+
+		context.contentResolver.query(
+			collection,
+			projection,
+			selection,
+			selectionArgs,
+			sortOrder,
+		)?.use { cursor ->
+			val idColumn = cursor.getColumnIndexOrThrow(MediaStore.Images.Media._ID)
+			val dateTakenColumn = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATE_TAKEN)
+
+			while (cursor.moveToNext()) {
+				val id = cursor.getLong(idColumn)
+				val dateTaken = cursor.getLong(dateTakenColumn).let {
+					if (it > 0) Date(it) else null
+				}
+
+				val contentUri = ContentUris.withAppendedId(
+					MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
+					id,
+				)
+
+				photos.add(
+					PhotoInfo(
+						uri = contentUri.toString(),
+						dateTaken = dateTaken,
+					),
+				)
+			}
+		}
+
+		photos
+	}
+}
```
저장소 내부의 사진 같은 컨텐츠는 contentResolver 를 이용해 불러올 수 있고, 쿼리 설정도 가능한가 보네요~

shinythink replied:
네 그렇습니다! sql을 kotlin 코드로 짜는 게 재미있네요

Ameri-Kano reviewed (approved):
너무 수고 많으셨습니다! 🙏
